lib/externalGraderLocal.js

```diff
@@ -97,7 +97,10 @@ class Grader {
                     },
                     Entrypoint: question.external_grading_entrypoint.split(' '),
                 }, (err, container) => {
-                    if (ERR(err, callback)) return;
+                    if (ERR(err, () => {})) {
+                        err.message = `Unable to launch Docker container for grading: ${err.message}`;
+                        callback(err);
```

Comment:
The problem here is that `callback()` is being called twice when there is an error, once on the `callback(err)` line and then it falls out of the `if` and calls `callback(null, container)` below. Calling the same `callback` twice in NodeJS gives the dreaded `Callback was already called` error 😄 

See https://nodejs.org/en/knowledge/errors/what-are-the-error-conventions/ for background on the callback convention used in NodeJS.

To fix this, we should (almost) always `return` after calling `callback()` as a safety mechanism to prevent further execution.

```suggestion
                        return callback(err);
```

Another possibility, which might be cleaner, would be to write:
```
if (err) err.message = `Unable to launch Docker container for grading: ${err.message}`;
if (ERR(err, callback)) return;
callback(null, container);
```

This second approach uses the standard `ERR()` pattern: see https://prairielearn.readthedocs.io/en/latest/dev-guide/#stack-traces-with-callback-style-functions

I think I originally suggested the pattern `if (ERR(err, () => {}))` in this case. I now regret this suggestion, sorry! While it does work fine, it also seems unnecessarily complex.

Reply:
Cool, tried both and they both worked -- currently using the latter approach since it's the readthedocs-endorsed method.